### PR TITLE
Mesh_3: connected components for 3D images

### PR DESCRIPTION
**Pull request by lrineau**
This PR adds an example to Mesh_3 showing how to initialize all connected components of a 3D image.

TODO:
- [x] Use the same in the `mesh_3_plugin` of the demo,
- [x] Add the documentation in the user manual. I am waiting for #1097 to be merged, to avoid conflicts.

@sloriot: the PR is not finished, but I would like it to be tested.

### CODE REVIEW

**Comment by sloriot** [comment]
`include\CGAL/Mesh_3/search_for_connected_components_in_labeled_image.h(111): warning C4267: 'argument': conversion from 'size_t' to 'const unsigned short', possible loss of data`
`include\CGAL/Mesh_3/search_for_connected_components_in_labeled_image.h(192): warning C4244: 'argument': conversion from 'const ptrdiff_t' to 'const unsigned short', possible loss of data`

**Comment by sloriot** [comment]
On all plaftorms:

```
------------------------------------------------------------------
- ProgramOutput.mesh_3D_image_with_custom_initialization.ArchLinux-CXX14
------------------------------------------------------------------
CGAL::get_default_random()::get_seed() = 1464452632
Searching for connected components...
  51 components were found.
Construct initial points...
  1006 initial points.
Warning /mnt/testsuite/include/CGAL/Mesh_3/Dump_c3t3.h:60
  the c3t3 object of following type:
N4CGAL33Mesh_complex_3_in_triangulation_3INS_23Regular_triangulation_3INS_46Robust_weighted_circumcenter_filtered_traits_3INS_5EpickEEENS_30Triangulation_data_structure_3INS_18Mesh_vertex_base_3IS4_NS_27Labeled_image_mesh_domain_3INS_7Image_3ES3_hiNS_7DefaultES9_EENS_27Triangulation_vertex_base_3IS4_NS_30Triangulation_ds_vertex_base_3IvEEEEEENS_24Compact_mesh_cell_base_3IS4_SA_vEENS_14Sequential_tagEEES9_EEiiEE
  cannot be dumped because some types are not streamable:
     - C3t3::Surface_patch_index is not streamable
       St4pairIiiE
```

**Comment by lrineau** [comment]
I have pushed a new version.

**Comment by lrineau** [comment]
The documentation is now in the PR, and was tested yesterday. @janetournois, @afabri, can you please comment (on the content, and on the English wordings).

[Here is the documentation, from CGAL-4.9-Ic-75, section "3.4.1Domains from Segmented 3D Images, with a Custom Initialization"][*].

[*]: https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-4.9-Ic-75/output/Mesh_3/index.html#Mesh_3DomainsFromSegmented3DImagesWithCustomInitialization

**Comment by janetournois** [comment]
I'm making a first pass. Andreas @afabri, you should wait for my commit to make yours

**Comment by janetournois** [comment]
it's ok for me

**Comment by janetournois** [comment]
for the other examples in the user manual, the code is displayed in a "box" in the text. Did you disable this on purpose or is it a mistake?

**Comment by lrineau** [comment]
... There is a box! For the section "3.4.1Domains from Segmented 3D Images, with a Custom Initialization", as well as for "3.6.3 Domains from Segmented 3D Images, with 1D Features", I have included only parts of examples, instead of the full `.cpp` file. But Doxygen uses a box in both cases. You do not see it?!

File: Mesh_3/doc/Mesh_3/Mesh_3.txt [hunk 1]
```diff
@@ -747,11 +747,11 @@ Right: the mesh generated after the initialization of all connected components
 
 Note that the example \ref
 Mesh_3/mesh_3D_image_with_custom_initialization.cpp also shows how to
-create an 3D image using the undocumente API of CGAL_ImageIO.
+create a 3D image using the undocumented API of CGAL_ImageIO.
 
 \snippet Mesh_3/mesh_3D_image_with_custom_initialization.cpp Create the image
 
-The code of the function `%random_labeled_image()` is in the example file \ref
+The code of the function `%random_labeled_image()` is in the header file \ref
```
Comment (lrineau) [inline Mesh_3/doc/Mesh_3/Mesh_3.txt]: Right, that is a header file, but it is in the `examples/Mesh_3` directory, and not in `include/CGAL`. That is why I named it an example file.

Comment (janetournois) [inline Mesh_3/doc/Mesh_3/Mesh_3.txt]: oh ok. So maybe "example header" to distinguish from "actual CGAL example"?

**Comment by janetournois** [comment]
I see multiple subboxes, which is probably enough/better, but it's different from the other examples, for which the complete example code is in a single big box

**Comment by lrineau** [comment]
Le Thursday 16 June 2016 03:59:33 Jane Tournois a écrit :
> I see multiple subboxes, which is probably enough/better, but it's different
> from the other examples, for which the complete example code is in a single
> big box

Indeed. Instead of the command `\cgalExample`, I have used the standard 
Doxygen commands `\snippet`, `\dontinclude`, `\skip`, and \until`, that allow 
to show parts of a file. The example file itself was either:
  - too big,
  - or too similar to the file `examples/Mesh_3/mesh_3D_image.cpp`.

I thought it was better that way: instead of a big blob of code, I have 
preferred to mix code lines with explanations.

-- 
Laurent Rineau, PhD
R&D Engineer at GeometryFactory           http://www.geometryfactory.com/
Release Manager of the CGAL Project       http://www.cgal.org/

**Comment by janetournois** [comment]
it's perfectly fine, I just wanted to mention it was different than other examples, and to check it was done on purpose

**Comment by lrineau** [comment]
I have also put that "feature" in the `mesh_3_plugin`. The PR request is finished.

At least once the review of the documentation is done. @afabri, can you have a look at the documentation. See https://github.com/CGAL/cgal/pull/1115#issuecomment-226426881. In particular, is that OK that the documentation only show a few lines of code, instead of showing the full .cpp file?

File: Mesh_3/test/Mesh_3/test_meshing_3D_image.cpp [hunk 2]
```diff
@@ -71,6 +71,9 @@ struct Image_tester : public Tester<K_e_i>
     // Verify
     this->verify_c3t3_volume(c3t3, 1772330*0.95, 1772330*1.05);
     this->verify(c3t3,domain,criteria, Bissection_tag());
+
+    typedef typename Mesh_domain::Surface_patch_index Patch_id;
+    CGAL_static_assertion(CGAL::Output_rep<Patch_id>::is_specialized);
```
Comment (sloriot) [inline Mesh_3/test/Mesh_3/test_meshing_3D_image.cpp]: You're missing a `CGAL_assertion_code` here

Comment (lrineau) [inline Mesh_3/test/Mesh_3/test_meshing_3D_image.cpp]: Fixed.

Comment (lrineau) [inline Mesh_3/test/Mesh_3/test_meshing_3D_image.cpp]: I have actually used `CGAL_USE_TYPE(..)` instead of `CGAL_assertion_code(..)`.

**Comment by sloriot** [comment]
@lrineau  are you still waiting for feedback? I no the branch can be merged IMO

**Comment by lrineau** [comment]
No. Both Andreas and Jane (@afabri and @janetournois) approved the documentation. I have already made the minor modifications after their feedback.

The PR can be merged.

**Pull request merged**